### PR TITLE
add more code to cloneset rollout plan

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
add more code to cloneset rollout plan
1. Add rollout plan mutation
2. Add validation
3. completed the rollout state transitions

### CODE REVIEW

File: apis/standard.oam.dev/v1alpha1/rollout_plan_types.go [hunk 1]
```diff
@@ -88,7 +87,10 @@ type RolloutPlan struct {
 	NumBatches *int32 `json:"numBatches,omitempty"`
 
 	// The exact distribution among batches.
-	// mutually exclusive to NumBatches
+	// mutually exclusive to NumBatches.
```
Comment (wonderflow) [inline apis/standard.oam.dev/v1alpha1/rollout_plan_types.go:90]: please mark a TODO here for webhook check.

File: pkg/controller/common/rollout/workloads/cloneset_controller.go [hunk 2]
```diff
@@ -148,37 +161,64 @@ func (c *CloneSetController) CheckOneBatchPods(ctx context.Context) *v1alpha1.Ro
...
-	c.recorder.Event(c.parentController, event.Normal("Finalized", "Rollout resource are finalized"))
+
 	c.rolloutStatus.StateTransition(v1alpha1.RollingFinalizedEvent)
+
 	return c.rolloutStatus
 }
 
-// The functions below are helper functions
+/* --------------------
+The functions below are helper functions
```
Comment (wonderflow) [inline pkg/controller/common/rollout/workloads/cloneset_controller.go:200]: why not use another file?

**Review by wonderflow — APPROVED** [review]
LGTM after nits

File: apis/standard.oam.dev/v1alpha1/rollout_plan_types.go (hunk 1 above)
Comment (ryanzhang-oss) [reply apis/standard.oam.dev/v1alpha1/rollout_plan_types.go:90]: I have already put the check there in this PR (I changed this comment while I am writing the check)

File: pkg/controller/common/rollout/workloads/cloneset_controller.go (hunk 2 above)
Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/workloads/cloneset_controller.go:200]: they are still methods (I've already separated out the utility functions). We could separate out all the helper functions if this file is too large in the end (it's almost complete now so it's not too big yet)

**Pull request merged**